Improve zooming with the mouse wheel

calixteman opened the pull request:
- Refactor the text layer code in order to avoid compute it on each draw: the idea is just to resuse what we got on the first draw. Now, we only update the scaleX of the different spans and the other values are dependant of --scale-factor.
- Move some properties in the CSS in order to avoid any updates in JS.
- When zooming in using mouse wheel, delay the page drawing and rescale in using css in order to minimize the drawing operations.
- with tracemonkey.pdf, we can see a real difference when profiling zooming operations (in using ctrl+wheel): there is nothing in the worker and this patch helps to remove all the jank markers.

Snuffleupagus commented on web/pdf_viewer.js:
```diff
@@ -1020,7 +1029,7 @@ class PDFViewer {
   }
 
-  #scrollIntoView(pageView, pageSpot = null) {
+  #scrollIntoView(pageView, timeout = null, pageSpot = null) {
```
All of these scroll-related changes makes me a fair bit worried, since delaying scrolling feels like something that could *potentially* cause other unforeseen issues.
Note that much of this code is quite old and there's lots of edge-cases to consider, so lots more comments explaining the logic would seem like a necessity here.

calixteman replied:
Ok, I'll see later how to improve that stuff.

Snuffleupagus reviewed (changes requested):
Please split this into a few different PRs, or at the very least a few different commits, since there's just way too many things changing all at once which feels *very* risky and thus really difficult to review.

*Edit:* In particular, implementing #15131 *separately* first would probably improve things a fair bit already.

calixteman committed:
Refactor the text layer code in order to avoid compute it on each draw

The idea is just to resuse what we got on the first draw.
Now, we only update the scaleX of the different spans and the other values
are dependant of --scale-factor.
Move some properties in the CSS in order to avoid any updates in JS.

Snuffleupagus commented on src/display/annotation_layer.js:
```diff
@@ -2659,29 +2658,12 @@ class AnnotationLayer {
-   * @param {HTMLDivElement} div
-   * @param {PageViewport} viewport
-   */
-  static #setDimensions(div, { width, height, rotation }) {
```
Here, and elsewhere, why are these methods being moved from the `src/`-folder and into the `web/`-folder? Won't that make the core PDF.js library less usable, assuming that we still care about non-Firefox use-cases here?

calixteman replied:
I changed the dimensions to something using the page dimensions and `--scale-factor` so now they're constant whatever we do (zooming, rotating, ...). So I thought it'd make more sense to have this code where the div properties are initialized.

calixteman committed:
Set the dimensions of the various layers at their creation

calixteman committed:
Delay the draw on rescaling with the mouse wheel and use css zoom

calixteman committed:
Don't delete the loadingIconDiv and reuse it

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -46,23 +47,43 @@ const DEFAULT_FONT_SIZE = 30;
+function getCtx(size, isOffscreenCanvasSupported, alpha = true) {
+  let ctx;
+  if (isOffscreenCanvasSupported && FeatureTest.isOffscreenCanvasSupported) {
+    ctx = new OffscreenCanvas(size, size).getContext("2d", { alpha });
```
Do we "need" to use `isOffscreenCanvasSupported` here, i.e. are the any clear benefits?
If not, maybe do that in a separate commit to reduce the size of these changes such that reviewing becomes slightly easier?

calixteman replied:
Please move your question here: https://github.com/mozilla/pdf.js/pull/15722 and I'll answer.

Snuffleupagus commented on web/pdf_viewer.css:
```diff
@@ -141,6 +141,10 @@
   display: none;
 }
 
+.pdfViewer .page .textLayer[hidden] {
+  display: none;
+}
+
```
All other layers simply do `div.hidden = true` in their respective `web/` JS-files, so can we please use the same pattern for this as well for consistency?

calixteman replied:
Same as above.

Snuffleupagus reviewed (commented):
As mentioned in the edit in https://github.com/mozilla/pdf.js/pull/15717#pullrequestreview-1188518849, can we please move the first commit (i.e. for addressing issue #15131) into its own PR?
